DylibBinder/DBFunc.cs

```diff
@@ -0,0 +1,133 @@
+			GenericParameters = dbProperty.GenericParameters;
+		}
+
+		public string Name { get; }
+		public bool IsStatic { get; }
+		public bool IsProperty { get; }
+		public string ReturnType { get; }
+		public bool HasThrows { get; }
+		public bool IsMutating { get; }
+		public bool HasInstance { get; }
+
+		public string PropertyType { get; } = string.Empty;
+		public bool IsPossiblyIncomplete { get; } = false;
+		public OperatorType OperatorKind { get; } = OperatorType.None;
+		public TypeAccessibility Accessibility { get; } = TypeAccessibility.Public;
+		public bool IsFinal { get; } = false;
+		public bool IsDeprecated { get; } = false;
+		public bool IsUnavailable { get; } = false;
+		public bool IsOptional { get; } = false;
+		public bool IsRequired { get; } = false;
+		public bool IsConvenienceInit { get; } = false;
+		public string ObjcSelector { get; } = string.Empty;
+
+		public DBParameterLists ParameterLists { get; }
+		public DBGenericParameters GenericParameters { get; }
+		public List<DBAssociatedType> AssociatedTypes { get; } = new List<DBAssociatedType> ();
+
+		bool GetStaticStatus (SwiftBaseFunctionType signature)
+		{
+			switch (signature) {
+			case SwiftStaticFunctionType:
+				return true;
+			case SwiftPropertyType propType:
+				return propType.IsStatic;
+			case SwiftUncurriedFunctionThunkType:
+				return false;
+			case SwiftUncurriedFunctionType:
+				return false;
+			default:
+				return false;
+			}
+		}
+	}
+
+	internal class DBFuncs {
+		public DBFuncs (ClassContents classContents)
+		{
+			var functions = SortedSetExtensions.CreateOverloadSortedSet ();
+			functions.AddRange (classContents.Methods.Values, classContents.StaticFunctions.Values, classContents.Constructors.Values);
+
+			foreach (var function in functions) {
+				foreach (var overloadFunction in function.Functions) {
+					if (overloadFunction.Name.Name.IsPublic () && !IsMetaClass (overloadFunction.Signature)
+						&& DoesNotDoubleThrow (overloadFunction.Signature.ToString ())) {
+						Funcs.Add (new DBFunc (overloadFunction));
+					}
+				}
+			}
+		}
+
+		public DBFuncs (ProtocolContents protocolContents)
+		{
+			var functions = SortedSetExtensions.CreateTLFunctionSortedSet ();
+			functions.AddRange (protocolContents.FunctionsOfUnknownDestination);
+			foreach (var function in functions) {
+				if (function.Name.Name.IsPublic () && !IsMetaClass (function.Signature)
+					&& DoesNotDoubleThrow (function.Signature.ToString ())) {
+					Funcs.Add (new DBFunc (function, isProtocol: true));
+				}
+			}
+		}
+
+		public List<DBFunc> Funcs { get; } = new List<DBFunc> ();
+
+		bool IsMetaClass (SwiftBaseFunctionType funcType)
+		{
+			if (funcType.ReturnType.Type == CoreCompoundType.MetaClass)
+				return true;
+			foreach (var param in funcType.EachParameter) {
+				if (param.Type == CoreCompoundType.MetaClass)
+					return true;
+			}
+			return false;
+		}
+
+		bool DoesNotDoubleThrow (string s)
+		{
+			var matches = Regex.Matches (s, "->");
```

Comment:
If there is more than one "->" in the string, I don't want to deal with it so I am using this to find out if there are multiple arrows in there. Is there a better way to do this?

Reply:
I actually have a larger problem with looking at the string representation of `function.Signature`. `ToString ()` was put in there strictly for debugging purposes and should not be used to determine the overall structure or functionality of the signature. The signature `TLFunction` is a `SwiftBaseFunctionType`. This is a decomposable/examinable type. You should be able to write the predicate you want based on that and if it makes sense, fold it into `SwiftBaseFunctionType`.
I also don't understand what it means to double throw. What does that mean?
As a final stylistic guideline, whenever possible, predicates should be written as a positive statement rather than a negative. "IsPurple" or "DoesThrow" or "ContainsArsenic".

Reply:
Discussed with Steve. I'm going to remove this DoesNotDoubleThrow method and the calls to it. This should have been named DoesNotRethrow but BTFS does not support this yet.